x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -332,6 +353,135 @@ private void createApiKeyAndIndexIt(
         }));
     }
 
+    public void updateApiKey(
+        Authentication authentication,
+        UpdateApiKeyRequest request,
+        Set<RoleDescriptor> userRoles,
+        ActionListener<UpdateApiKeyResponse> listener
+    ) {
+        ensureEnabled();
+
+        if (authentication == null) {
+            listener.onFailure(new IllegalArgumentException("authentication must be provided"));
+            return;
+        }
+
+        findVersionedApiKeyDocsForSubject(authentication, new String[] { request.getId() }, ActionListener.wrap((versionedDocs) -> {
+            final var apiKeyId = request.getId();
+
+            if (versionedDocs.isEmpty()) {
+                throw apiKeyNotFound(apiKeyId);
+            }
```

Comment:
Agreed that the error message is too generic and will be confusing.

I think a 404 can still be appropriate here though since we are looking for a resource by its ID with extra filter parameters of owner name and realm name and are not finding it.

I'm good with "API key not found or not owned by the requesting user". Alternatively, wondering about `API key owned by requesting user not found` since it more explicitly describes the resource we were unable to find?

I'd rather avoid the extra `GetApiKey` call -- it seems overkill just to construct a message.

Reply:
I am OK with just expanding the error message. I also like your 2nd version. Unrelated: between "API key" and "api key", what do you prefer? I personally prefer the former but we haven't been consistent.

Reply:
I'm good with the former! I picked lowercase at random as the first "style" I've seen but API should be capped.